### DISCUSSION TITLE
Recommended way to workaround IF [NOT] EXISTS

### DISCUSSION THREAD

**Discussion by dazhuangc** [sequelize/sequelize#17360] [Help & Questions] [answered]
I am now implementing HANA dialect.
HANA does not support `IF [NOT] EXISTS` like `CREATE TABLE IF NOT EXISTS "table1"` and `DROP TABLE IF EXISTS "table1"`.

The `IF [NOT] EXISTS` seems not an option of `createTable()` or `dropTable()`, so I think it is necessary to implement them in HANA dialect. 

I have 2 ideas to workaround:
1. Generate a complex SQL statement in `dropTableQuery()` in `query-generator-typescript.internal.ts`, like:
    ```sql
    DO BEGIN
      DECLARE table_count INTEGER;
      SELECT COUNT(*) INTO table_count FROM TABLES 
      WHERE TABLE_NAME = 'table1' AND SCHEMA_NAME = 'schema1';
      IF :table_count > 0 THEN
        EXEC 'DROP TABLE IF EXISTS "table1";
      END IF;
    END;
    ```
    And let `query-interface` execute this complex SQL statement.

2. Override `createTable()` in `query-interface-typescript.internal.ts`:
    ```typescript
      async createTable<M extends Model>(
        tableName: TableName,
        attributes: CreateTableAttributes<M, CreationAttributes<M>>,
        options?: QueryInterfaceCreateTableOptions,
      ): Promise<void> {
        const tableExists = await this.tableExists(tableName);
        if (!tableExists) {
          return super.createTable(tableName, attributes, options);
        }
      };
    ```
    Before executing SQL statements to create or drop table, `query-interface` needs to check if the table exists.

Which one do you think is better?

**Comment by WikiRik** [comment, marked as answer]
I'd prefer to add the check in queryGenerator. We want to release queryGenerator as a public API in the future and limit the amount of logic done in queryInterface.

**Reply by dazhuangc** [reply]
Thanks, then I will try to implement `IF [NOT] EXISTS` in queryGenerator.
